File: webresource/webstream.cpp
#include "webstream.h"

using namespace emscripten;

EMSCRIPTEN_BINDINGS(WebStreamFFmpeg){
    class_<WebStreamFFmpeg>("WebStreamFFmpeg")
       .constructor<>()
       .function("AddRef", &WebStreamFFmpeg::AddRef)
       .function("Release", &WebStreamFFmpeg::Release)
       .function("setStreamType", &WebStreamFFmpeg::setStreamType)
       .function("getStreamType", &WebStreamFFmpeg::getStreamType)
       .function("setStreamCodecType", &WebStreamFFmpeg::setStreamCodecType)
       .function("getStreamCodecType", &WebStreamFFmpeg::getStreamCodecType)
       .function("getDuration", &WebStreamFFmpeg::getDuration)
       .function("setDuration", &WebStreamFFmpeg::setDuration)
       .function("setStreamIndex", &WebStreamFFmpeg::setStreamIndex)
       .function("getStreamIndex", &WebStreamFFmpeg::getStreamIndex)
       .function("streamName", &WebStreamFFmpeg::streamName)
       .function("getStreamTitle", &WebStreamFFmpeg::getStreamTitle)
       .function("setStreamTitle", &WebStreamFFmpeg::setStreamTitle)
       .function("getFrameWidth", &WebStreamFFmpeg::getFrameWidth)
       .function("getFrameHeight", &WebStreamFFmpeg::getFrameHeight)
       .function("setFrameSize", &WebStreamFFmpeg::setFrameSize)
       .function("getFrameFormat", &WebStreamFFmpeg::getFrameFormat)
       .function("setFrameFormat", &WebStreamFFmpeg::setFrameFormat)
       .function("getAudioSampleRate", &WebStreamFFmpeg::getAudioSampleRate)
       .function("getAudioChannelCount", &WebStreamFFmpeg::getAudioChannelCount)
       .function("getAudioSampleFormat", &WebStreamFFmpeg::getAudioSampleFormat)
       .function("setAudioFormat", &WebStreamFFmpeg::setAudioFormat)
       .function("getTimebase", &WebStreamFFmpeg::getTimebase)
       .function("setTimebase", &WebStreamFFmpeg::setTimebase)
       .function("getFrameRate", &WebStreamFFmpeg::getFrameRate)
       .function("setFrameRate", &WebStreamFFmpeg::setFrameRate)
       .function("getCodecExtradataMem", &WebStreamFFmpeg::getCodecExtradataMem)
       .function("setCodecExtradata", &WebStreamFFmpeg::setCodecExtradata)

       ;
}

EMSCRIPTEN_BINDINGS(StreamAVMediaType) {
    enum_<IAVStream::AVMediaType>("StreamAVMediaType")
        .value("AVTypeUnknow", IAVStream::AVTypeUnknow)
        .value("AVTypeVideo", IAVStream::AVTypeVideo)
        .value("AVTypeAudio", IAVStream::AVTypeAudio)
        .value("AVTypeData", IAVStream::AVTypeData)
        .value("AVTypeSubtitle", IAVStream::AVTypeSubtitle)
        .value("AVTypeAttachment", IAVStream::AVTypeAttachment)
        ;
}

EMSCRIPTEN_BINDINGS(StreamAudioSampleFormat) {
    enum_<IAVStream::AudioSampleFormat>("StreamAudioSampleFormat")
        .value("AudioFormateNone", IAVStream::AudioFormateNone)
        .value("AudioFormateU8", IAVStream::AudioFormateU8)
        .value("AudioFormateS16", IAVStream::AudioFormateS16)
        .value("AudioFormateS32", IAVStream::AudioFormateS32)
        .value("AudioFormateFLT", IAVStream::AudioFormateFLT)
        .value("AudioFormateDBL", IAVStream::AudioFormateDBL)
        .value("AudioFormateU8P", IAVStream::AudioFormateU8P)
        .value("AudioFormateS16P", IAVStream::AudioFormateS16P)
        .value("AudioFormateS32P", IAVStream::AudioFormateS32P)
        .value("AudioFormateFLTP", IAVStream::AudioFormateFLTP)
        .value("AudioFormateDBLP", IAVStream::AudioFormateDBLP)
        .value("AudioFormateS64", IAVStream::AudioFormateS64)
        .value("AudioFormateS64P", IAVStream::AudioFormateS64P)
        ;
}

EMSCRIPTEN_BINDINGS(StreamAVCodecType) {
    enum_<IAVStream::AVCodecType>("StreamAVCodecType")
        .value("CodecTypeUnkow", IAVStream::CodecTypeUnkow)
        .value("CodecTypeH264", IAVStream::CodecTypeH264)
        .value("CodecTypeHevc", IAVStream::CodecTypeHevc)
        .value("CodecTypeAAC", IAVStream::CodecTypeAAC)
        .value("CodecTypePCMMULAM", IAVStream::CodecTypePCMMULAM)
        .value("CodecTypePCMALAW", IAVStream::CodecTypePCMALAW)
        ;
}

EMSCRIPTEN_BINDINGS(StreamFrameFormat) {
    enum_<IAVStream::FrameFormat>("StreamFrameFormat")
        .value("FormatNone", IAVStream::FormatNone)
        .value("FormatRGB24", IAVStream::FormatRGB24)
        .value("FormatRGBA", IAVStream::FormatRGBA)
        .value("FormatBGRA", IAVStream::FormatBGRA)
        .value("FormatGRAY8", IAVStream::FormatGRAY8)
        .value("FormatYUV420P", IAVStream::FormatYUV420P)
        .value("ForamtARGB", IAVStream::ForamtARGB)
        .value("ForamtNV12", IAVStream::ForamtNV12)
        .value("ForamtNV21", IAVStream::ForamtNV21)
        ;
}

EMSCRIPTEN_BINDINGS(StreamRational){
    value_array<IAVStream::StreamRational>("StreamRational")
       .element(&IAVStream::StreamRational::num)
       .element(&IAVStream::StreamRational::den)
       ;
}

File: webplayer/h5player/resource/WebStream.js
class WebStream {
    constructor(stream) {
        if(stream == null){
            this.duration = 0;
            this.streamTitle = "";
            this.frameWidth = 0;
            this.frameHeight = 0;
            this.frameFormat = -1;
            this.nSampleRate = 0;
            this.nChannelCount = 0;
            this.audioSampleFormat = -1;
            this.timebase = [1,1000];
            this.frameRate = [1,1];
            return;
        }
        this.streamType = stream.getStreamType().value;
        this.streamCodecType = stream.getStreamCodecType().value;
        this.duration = stream.getDuration();
        this.streamIndex = stream.getStreamIndex();  
        this.streamName= stream.streamName(); 
        this.streamTitle = stream.getStreamTitle();

        this.frameWidth = stream.getFrameWidth();
        this.frameHeight = stream.getFrameHeight();

        this.frameFormat = stream.getFrameFormat().value;

        this.nSampleRate = stream.getAudioSampleRate();
        this.nChannelCount = stream.getAudioChannelCount();
        this.audioSampleFormat = stream.getAudioSampleFormat().value;

        this.timebase = stream.getTimebase();
        this.frameRate = stream.getFrameRate();

        let codecExtradata = stream.getCodecExtradataMem();
        this.codecExtradata = new Uint8Array(codecExtradata.length);
        this.codecExtradata.set(codecExtradata, 0);
    }

    convertToStreamFFmpeg(streamFFmpeg) {
        switch (this.streamType) {
            case -1:
                this.streamType = Module.StreamAVMediaType.AVTypeUnknow;
                break;
            case 0:
                this.streamType = Module.StreamAVMediaType.AVTypeVideo;
                break;
            case 1:
                this.streamType = Module.StreamAVMediaType.AVTypeAudio;
                break;
            case 2:
                this.streamType = Module.StreamAVMediaType.AVTypeData;
                break;
            case 3:
                this.streamType = Module.StreamAVMediaType.AVTypeSubtitle;
                break;
            case 4:
                this.streamType = Module.StreamAVMediaType.AVTypeAttachment;
                break;
            default:
                this.streamType = Module.StreamAVMediaType.AVTypeUnknow;
        }

        streamFFmpeg.setStreamType(this.streamType);

        switch (this.streamCodecType) {
            case -1:
                this.streamCodecType = Module.StreamAVCodecType.CodecTypeUnkow;
                break;
            case 0:
                this.streamCodecType = Module.StreamAVCodecType.CodecTypeH264;
                break;
            case 1:
                this.streamCodecType = Module.StreamAVCodecType.CodecTypeHevc;
                break;
            case 2:
                this.streamCodecType = Module.StreamAVCodecType.CodecTypeAAC;
                break;
            case 3:
                this.streamCodecType = Module.StreamAVCodecType.CodecTypePCMMULAM;
                break;
            case 4:
                this.streamCodecType = Module.StreamAVCodecType.CodecTypePCMALAW;
                break;
            default:
                this.streamCodecType = Module.StreamAVCodecType.CodecTypeUnkow;
                break;
        }
        streamFFmpeg.setStreamCodecType(this.streamCodecType);

        streamFFmpeg.setDuration(this.duration);
        streamFFmpeg.setStreamIndex(this.streamIndex);
        streamFFmpeg.setStreamTitle(this.streamTitle);


        switch (this.frameFormat) {
            case -1:
                this.streamCodecType = Module.StreamFrameFormat.FormatNone;
                break;
            case 0:
                this.streamCodecType = Module.StreamFrameFormat.FormatRGB24;
                break;
            case 1:
                this.streamCodecType = Module.StreamFrameFormat.FormatRGBA;
                break;
            case 2:
                this.streamCodecType = Module.StreamFrameFormat.FormatBGRA;
                break;
            case 3:
                this.streamCodecType = Module.StreamFrameFormat.FormatGRAY8;
                break;
            case 4:
                this.streamCodecType = Module.StreamFrameFormat.FormatYUV420P;
                break;
            case 5:
                this.streamCodecType = Module.StreamFrameFormat.ForamtARGB;
                break;
            case 6:
                this.streamCodecType = Module.StreamFrameFormat.ForamtNV12;
                break;
            case 7:
                this.streamCodecType = Module.StreamFrameFormat.ForamtNV12;
                break;
            default:
                this.streamCodecType = Module.StreamFrameFormat.FormatNone;
                break;
        }

        streamFFmpeg.setFrameFormat(this.frameFormat);

        streamFFmpeg.setFrameSize(this.frameWidth,this.frameHeight);

        switch(this.audioSampleFormat) {
            case -1:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateNone;
                break;
            case 0:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateU8;
                break;
            case 1:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateS16;
                break;
            case 2:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateS32;
                break;
            case 3:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateFLT;
                break;
            case 4:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateDBL;
                break;
            case 5:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateU8P;
                break;
            case 6:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateS16P;
                break;
            case 7:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateS32P;
                break;
            case 8:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateFLTP;
                break;
            case 9:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateDBLP;
                break;
            case 10:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateS64;
                break;
            case 11:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateS64P;
                break;    
            default:
                this.audioSampleFormat = Module.StreamAudioSampleFormat.AudioFormateNone;
            }
        streamFFmpeg.setAudioFormat(this.nSampleRate,this.nChannelCount,this.audioSampleFormat);

        streamFFmpeg.setTimebase(this.timebase);
        streamFFmpeg.setFrameRate(this.frameRate);

        streamFFmpeg.setCodecExtradata(this.codecExtradata);
    }
}